### ISSUE TITLE
Negate mutator

### ISSUE THREAD

**Issue by richardwerkman** [stryker-mutator/stryker-net#444]
Currently Stryker mutates expressions from negative to non negative [here](https://github.com/stryker-mutator/stryker-net/blob/8b1b900513edcf55e07cafbf2824d7d4ecaae091/src/Stryker.Core/Stryker.Core/Mutators/PrefixUnaryMutator.cs#L21). But not the other way around. 

![image](https://user-images.githubusercontent.com/14224500/55628620-7f6a1f80-57b1-11e9-997d-1c86ba2fbe33.png)
This statement is not being mutated by stryker. It could be negated.

## Why?
It would be nice to see stryker mutate this expression into:
``` cs
if (!scan.MoveNext()) {
```

## The catch
We have to prevent double mutations from happening. For example:
![image](https://user-images.githubusercontent.com/14224500/55628698-ac1e3700-57b1-11e9-947d-8a9a24b8ca34.png)
Should not be mutated more than once. 

This means the negate mutator should specificly target expressions that do not contain any other mutations. This is hard to detect since our mutators have no knowledge of other mutations. We have to think of specific SyntaxNodes that should be mutated.

**Comment by dupdob** [comment]
It may be simple: if we cannot generate mutant for the condition then Stryker add the negated mutant

It should do the trick

Bearing in mind that there is a method dedicated to if statement mutations..

Envoyé de mon iPhone

**Comment by richardwerkman** [comment]
@dupdob don't we want to mutate more than if statements? I can imagine this is useful in where expressions too. Or even in method call arguments. 

We could state that every expression that contains a method call without any arguments should be negated. That way we can create a regular mutator and don't have to change the mutantorchestrator. What do you think?

**Comment by dupdob** [comment]
@richardwerkman : well, that sounds hard to me: this mutator works only on boolean expression. The only places where we can be sure of that are:
- conditions within statements (if, for, while)
- boolean literals (true, false)
you add to that the logic that requires mutating expressions that cant be mutated in any other way..

**Comment by dupdob** [comment]
Let me push an in interesting question: how shall we mutate this:

```csharp
if (Check(true))
{
...
}
```
does it make sens to have both:
```csharp
if (!Check(false))
{
...
}
```
and
```csharp
if (Check(true))
{
...
}
```

Hence my proposal to focus on *if, while, for (maybe)*, at least for a first try. 
That being said, you may look at other mutation framework to see their strategies for this situation.

**Comment by richardwerkman** [comment]
Well, in my proposal I said: Method calls with no arguments. So 
``` cs
if (Check(true))
{
...
}
```
Would not be mutated as it has an argument.
``` cs
if (Check())
{
...
}
```
Would be mutated as it has no arguments and thus would not be mutated anyway. 

We could just target `InvocationExpression` and check if the argumentlist is empty.

I'm now thinking, we could even check if the parent node of the Expression is an If statement or While statement. If it's an other node we could skip it from negating. 

My preference goes to putting all logic inside the mutator and not in the mutatorOrchestrator. The mutatorOrchestrator already is a mess, so I'd like to keep it as it is.

**Comment by dupdob** [comment]
oups, my examples were inverted, sorry for the mishap.
ok, your answer makes sense. I agree we need to delegate the responsibility to the mutator.
The question is then, should we:
(1) let the mutator discover the context,
(2) add a parameter to provide hints to the context?

(1) provides maximum freedom, at the cost of performance and risk of mutator code getting ugly (as seen in mutant orchestrator)
(2) faster and cleaner, but require some refactoring when new informations are required.

I have a bias towards (2) because: mutation (process) performance is important and I prefer to avoid duplicate complexity due to context (re)discovery.

But, as usual, your call, guys

**Comment by richardwerkman** [comment]
Closed by #451
